lib/Sema/CSSolver.cpp

```diff
@@ -1305,8 +1305,10 @@ ConstraintSystem::filterDisjunction(
   for (unsigned constraintIdx : indices(disjunction->getNestedConstraints())) {
     auto constraint = disjunction->getNestedConstraints()[constraintIdx];
 
-    // Skip already-disabled constraints.
-    if (constraint->isDisabled())
+    // Skip already-disabled constraints. Let's treat disabled
+    // choices which have a fix as "enabled" ones here, so we can
+    // potentially infer some type information from them.
+    if (constraint->isDisabled() && !constraint->getFix())
```

Comment:
There's only one client of `filterDisjunction` right now, and it will be less able to break a system into more components if it looks at disabled constraints that have fixes. For example, given something like this:

```swift
func f(a: Int) -> Int { ... }
func f(a: Double) -> Int { ... }
func f(b: Double) -> Double { ... }

z + f(a: x)
```

In the call at the end, the disjunction choice for `f(b:)` will be disabled, presumably with a fix to change the label.  Currently, that means only the two `f(a:)` overloads will be considered, so the "common result type" of the call can be determined to be `Int`, (which can be used to help solve further without having to resolve the overload). However, if we start including `f(b:)` in the computation, we no longer have a "common result type" and we lose out on the solver optimization.

Reply:
I think labeling changes should be decided after overload choices are in the list, so it would be completely up to `filterDisjunction` to decide whether to attach fixes to some of the choices... That's why I want to remove `UR_LabelMismatch` from lookup result. Currently more structural errors are supported e.g. referencing instance members on metatype base, invalid partial application and soon inaccessible members.

Also I think what you are describing is more of a problem with `filterDisjunction` optimization than these changes, because it is too eager when it comes to "fixable" constraints. But I don't think there is anything wrong with including candidates which could possibly match.

That said it's might be possible to avoid considering them in `filterDisjunction` but that would take a separate PR to do, that's why I've included them instead of trying to re-design the whole thing in this PR, here are a couple of problem areas:

- If there is only one choice which meets the criteria - disjunction is dissolved, which is a wrong thing to do in presence of fixable overloads, we should probably retire the disjunction and apply the overload instead so it could be resurrected later.

Here is the example for this problem:

```swift
struct X { }
struct Y { }

class A {
  func mixed(x: X) -> X { return X() }
  class func mixed(y: Y) -> Y { return Y() }
}

func foo(_ a: A, y: Y) {
  var y1 = a.mixed(y: y) // filterDisjunction applies `mixed(x:)` and discards `mixed(y:)`
}
```

- Common result type uses `bind constraint` which means that, if it's computed incorrectly, that would prevent other choices to be attempted, which impedes diagnostics e.g. previous example but with added contextual type to `y1` e.g. `var y1: Y = a.mixed(y: y)`. Even if `mixed(y:)` has left in the list it would still not match because "common type" would be deduced to be `X`.

Reply:
The fundamental issue here, I think, is that we're halfway between "ignore disabled constraints so we can solve the system faster" and "don't make decisions that contradict what's in those disabled constraints, because then we can't correctly consider those when generating diagnostics". There is no easy answer here---the right answer is to delay any computation that involves looking at disabled constraints until after we've set up a solver scope, so we can introduce a scope where we ignore all the disabled constraints (to get to a solution fastest for well-formed code, but can still back out of that if things go badly.

The common result type computation (as a solver step) is new enough that perhaps we won't see a measurable effect from considering disabled cases with fixes in it.

Reply:
I agree, that's what I tried to convey in my previous comment! I don't think there is big enough impact on well-formed code from these changes and not sure I can address this problem in this PR.

Reply:
Alright. Please go ahead and merge. As we move more optimizations into the solver proper, we'll re-evaluate.